R/transformCounts.R

```diff
@@ -307,6 +346,120 @@ setMethod("transformAssay", signature = c(x = "SummarizedExperiment"),
+    # constant in additional args
+    if (is.null(css_percentile)) {
+      css_percentile <- ifelse("css_percentile" %in% names(list(...)), 
+                               list(...)$css_percentile, 0.5)
+    }
+    scaling_constant <- ifelse("scaling_constant" %in% names(list(...)), 
+                               list(...)$scaling_constant, 1000)
+    
+    # Call .calc_scaling_factors method to calculate scaling factors
+    scaling_factors <- .calc_scaling_factors(mat, css_percentile)
+    
+    # Normalize the count data by dividing by the scaling factor
+    normalized_data <- sweep(mat, 2, scaling_factors / scaling_constant, "/")
+    return(normalized_data)
+}
+
+####################################.calc_css_fast##############################
+# This function applies cumulative sum scaling (CSS) to the abundance table 
+# using the fast variant method.
+.calc_css_fast <- function(mat, ...) {
+    # Check for the presence of the rel additional arg and set default if not present
+    rel <- ifelse("rel" %in% names(list(...)), list(...)$rel, 0.1)
+    
+    # Calculate the percentile using .calc_css_percentile
+    css_percentile <- .calc_css_percentile(mat, rel, ...)
+    
+    # Call .calc_css with the calculated percentile
+    normalized_data <- .calc_css(mat, css_percentile, ...)
+    return(normalized_data)
+}
+
+#################################.calc_css_percentile######################
+# Calculates the cumulative sum scaling (css) scaling percentiles from the given 
+# data
+#'  @importFrom matrixStats rowMedians
+.calc_css_percentile <- function(mat, rel, ...) {
+    # Sort the non-zero values in each column of the matrix in decreasing order
+    smat <- lapply(1:ncol(mat), function(i) {
+      sort(mat[which(mat[, i] > 0), i], decreasing = TRUE)
+    })
+    
+    # Find the maximum length of the sorted columns
+    leng <- max(sapply(smat, length))
+    
+    # Check if any column has only one or zero features and stop if true
+    if (any(sapply(smat, length) == 1)) stop("Warning: sample with one or zero features")
+    
+    # Create an empty matrix (smat2) with dimensions (max length, number of columns)
+    smat2 <- array(NA, dim = c(leng, ncol(mat)))
+    
+    # Populate smat2 with the sorted values from smat
+    for (i in 1:ncol(mat)) {
+      smat2[leng:(leng - length(smat[[i]]) + 1), i] <- smat[[i]]
+    }
+    
+    # Calculate quantiles for each column in smat2
+    rmat2 <- sapply(1:ncol(smat2), function(i) {
+      quantile(smat2[, i], probs = seq(0, 1, length.out = nrow(smat2)), na.rm = TRUE)
+    })
+    
+    # Replace NA values in smat2 with 0
+    smat2[is.na(smat2)] <- 0
+    
+    # Compute the row means of smat2 to create a reference profile
+    ref1 <- rowMeans(smat2)
+    
+    # Number of columns in rmat2
+    ncols <- ncol(rmat2)
+    
+    # Calculate the differences between the reference profile and each column's quantiles
+    diffr <- sapply(1:ncols, function(i) {
```

Comment:
Instead of 1:10, use seq_len(10)

Reply:
Also, based on Bioconductor recommendations, avoid sapply(). Use vapply or lapply